libraries/LittleFS/src/LittleFS.cpp

```diff
@@ -0,0 +1,133 @@
+        return FileImplPtr();
+    }
+    return std::make_shared<LittleFSFileImpl>(this, path, fd);
+}
+
+DirImplPtr LittleFSImpl::openDir(const char *path) {
+    if (!_mounted) {
+        return DirImplPtr();
+    }
+    auto dir = std::make_shared<lfs_dir_t>();
+    int rc = lfs_dir_open(&_lfs, dir.get(), "/");
+    if (rc < 0) {
+        DEBUGV("LittleFSImpl::openDir: path=`%s` err=%d\n", path, rc);
+        return DirImplPtr();
+    }
+    // Skip the . and .. entries
+    lfs_info dirent;
+    lfs_dir_read(&_lfs, dir.get(), &dirent);
+    lfs_dir_read(&_lfs, dir.get(), &dirent);
+
+    return std::make_shared<LittleFSDirImpl>(path[0]=='/'?(path+1):path, this, dir);
+}
+
+
+// Borrow the existing SPIFFS flash HAL
+extern int32_t spiffs_hal_write(uint32_t addr, uint32_t size, const uint8_t *src);
+extern int32_t spiffs_hal_erase(uint32_t addr, uint32_t size);
+extern int32_t spiffs_hal_read(uint32_t addr, uint32_t size, uint8_t *dst);
+
+int LittleFSImpl::lfs_flash_read(const struct lfs_config *c,
+    lfs_block_t block, lfs_off_t off, void *dst, lfs_size_t size) {
+    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
+    uint32_t addr = me->_start + (block * me->_blockSize) + off;
+    return spiffs_hal_read(addr, size, static_cast<uint8_t*>(dst));
+}
+
+int LittleFSImpl::lfs_flash_prog(const struct lfs_config *c,
+    lfs_block_t block, lfs_off_t off, const void *buffer, lfs_size_t size) {
+    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
+    uint32_t addr = me->_start + (block * me->_blockSize) + off;
+    const uint8_t *src = reinterpret_cast<const uint8_t *>(buffer);
+    return spiffs_hal_write(addr, size, static_cast<const uint8_t*>(src));
+}
+
+int LittleFSImpl::lfs_flash_erase(const struct lfs_config *c, lfs_block_t block) {
+    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
+    uint32_t addr = me->_start + (block * me->_blockSize);
+    uint32_t size = me->_blockSize;
+    return spiffs_hal_erase(addr, size);
+}
+
+int LittleFSImpl::lfs_flash_sync(const struct lfs_config *c) {
+    /* NOOP */
+    (void) c;
+    return 0;
+}
+
+
+// these symbols should be defined in the linker script for each flash layout
+#ifndef CORE_MOCK
+#ifdef ARDUINO
+#ifndef SPIFFS_MAX_OPEN_FILES
+#define SPIFFS_MAX_OPEN_FILES 5
+#endif
+
+#ifdef ARDUINO
+extern "C" uint32_t _SPIFFS_start;
+extern "C" uint32_t _SPIFFS_end;
+extern "C" uint32_t _SPIFFS_page;
+extern "C" uint32_t _SPIFFS_block;
+
+#define SPIFFS_PHYS_ADDR ((uint32_t) (&_SPIFFS_start) - 0x40200000)
+#define SPIFFS_PHYS_SIZE ((uint32_t) (&_SPIFFS_end) - (uint32_t) (&_SPIFFS_start))
+#define SPIFFS_PHYS_PAGE ((uint32_t) &_SPIFFS_page)
+#define SPIFFS_PHYS_BLOCK ((uint32_t) &_SPIFFS_block)
+#endif
+
+#if !defined(NO_GLOBAL_INSTANCES) && !defined(NO_GLOBAL_LITTLEFS)
+FS LittleFS = FS(FSImplPtr(new LittleFSImpl( SPIFFS_PHYS_ADDR, SPIFFS_PHYS_SIZE, SPIFFS_PHYS_PAGE, SPIFFS_PHYS_BLOCK, SPIFFS_MAX_OPEN_FILES)));
```

Comment:
With the namespace comment for the .h file, this would become:
```
FS LittleFS = FS(FSImplPtr(new LittleFSImplementation::LittleFSImpl( SPIFFS_PHYS_ADDR,....)
```

Reply:
Done latest commit